stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -243,34 +238,151 @@ pub fn get_nakamoto_reward_cycle_info<U: RewardSetProvider>(
+    };
+
+    info!(
+        "Anchor block selected";
+        "cycle" => reward_cycle,
+        "block_id" => %block_id,
+        "consensus_hash" => %anchor_block_header.consensus_hash,
+        "burn_height" => anchor_block_header.burn_header_height,
+        "anchor_chain_tip" => %anchor_block_header.index_block_hash(),
+        "anchor_chain_tip_height" => %anchor_block_header.burn_header_height,
+    );
+
+    return Ok(Some(rc_info));
+}
+
+/// Helper to get the Nakamoto reward set for a given reward cycle.
+/// In all but the first Nakamoto reward cycle, this will load up the stored reward set from the
+/// Nakamoto chain state.  In the first Nakamoto reward cycle, where the reward set is computed
+/// from epoch2 state, the reward set will be loaded from the sortition DB (which is the only place
+/// it will be stored).
+///
+/// Returns Ok(Some((reward set info, PoX anchor block header))) on success
+/// Returns Ok(None) if the reward set is not yet known, but could be known by the time a
+/// subsequent call is made.
+pub fn load_nakamoto_reward_set<U: RewardSetProvider>(
+    burn_height: u64,
+    sortition_tip: &SortitionId,
+    burnchain: &Burnchain,
+    chain_state: &mut StacksChainState,
+    sort_db: &SortitionDB,
+    provider: &U,
+) -> Result<Option<(RewardCycleInfo, StacksHeaderInfo)>, Error> {
+    let epoch_at_height = SortitionDB::get_stacks_epoch(sort_db.conn(), burn_height)?
+        .unwrap_or_else(|| panic!("FATAL: no epoch defined for burn height {}", burn_height));
+
+    // calculating the reward set for the _next_ reward cycle
+    let reward_cycle = burnchain
+        .next_reward_cycle(burn_height)
+        .expect("FATAL: no reward cycle for burn height");
+
+    let reward_start_height = burnchain.reward_cycle_to_block_height(reward_cycle);
+
     // Find the first Stacks block in this reward cycle's preceding prepare phase.
     // This block will have invoked `.signers.stackerdb-set-signer-slots()` with the reward set.
     // Note that we may not have processed it yet. But, if we do find it, then it's
     // unique (and since Nakamoto Stacks blocks are processed in order, the anchor block
     // cannot change later).
-    let prepare_phase_sortitions =
-        find_prepare_phase_sortitions(sort_db, burnchain, sortition_tip)?;
-
-    // did we already calculate the reward cycle info?  If so, then return it.
-    let first_sortition_id = if let Some(first_sn) = prepare_phase_sortitions.first() {
-        if let Some(persisted_reward_cycle_info) =
-            SortitionDB::get_preprocessed_reward_set(sort_db.conn(), &first_sn.sortition_id)?
+    let first_epoch30_reward_cycle = burnchain
+        .next_reward_cycle(epoch_at_height.start_height)
+        .expect("FATAL: no reward cycle for epoch 3.0 start height");
+
+    if epoch_at_height.epoch_id < StacksEpochId::Epoch30
+        || (epoch_at_height.epoch_id == StacksEpochId::Epoch30
+            && reward_cycle == first_epoch30_reward_cycle)
```

Comment:
It'd be better for this epoch gating to be implemented via `epoch_id.uses_nakamoto_reward_set()` or similar. That way, the epoch "rules" all live in the `StacksEpochId` impl.

Reply:
Yup, can do